sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -783,6 +783,16 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       ds.filter(_.b > 1).collect().toSeq
     }
   }
+
+  test("transformed dataset correctly solve the attributes") {
+    val dataset = Seq(1, 2, 3).toDS()
+    val ds1 = dataset.map(_ + 1)
```

Comment:
nit: I'd like it to be:
```
val ds = Seq(1, 2, 3).toDS().map(_ + 1)
val d1 = ds.as("d1")
val d2 = ds.as("d2")
...
```

and the test name can be: `mapped dataset should resolve duplicated attributes for self join and set operations`